crates/flipperzero/src/notification/mod.rs

```diff
@@ -0,0 +1,308 @@
+//! Furi notifications.
+
+use core::ffi::c_char;
+
+use flipperzero_sys as sys;
+use flipperzero_sys::furi::UnsafeRecord;
+
+///Default notification messages.
+pub mod messages;
+///Default notification notes.
+pub mod notes;
+///Default notification sequences.
+pub mod sequences;
+
+const RECORD_NOTIFICATION: *const c_char = sys::c_string!("notification");
+
+/// A handle to the Notification app.
+pub struct NotificationApp {
+    data: UnsafeRecord<sys::NotificationApp>,
+}
+
+impl NotificationApp {
+    /// Obtains a handle to the Notifications app.
+    pub fn open() -> Self {
+        Self {
+            data: unsafe { UnsafeRecord::open(RECORD_NOTIFICATION) },
+        }
+    }
+
+    /// Runs a notification sequence.
+    ///
+    /// #Safety
+    /// Due to how rust interacts with the firmware this function is not safe to use at any time
+    /// where the application might exit directly afterwards as the rust runtime will free the
+    /// sequence before the firmware has finished reading it. At any time where this is an issue
+    /// `notify_blocking` should be used instead..
```

Comment:
Good news! This was resolved in https://github.com/flipperdevices/flipperzero-firmware/pull/2335.

Reply:
Great! Do you think this may allow us to use dynamically allocated sequences? I would love to test myself but I'm honestly not sure on how to sync this branch with `main`, I've only ever done so in my own repos where I'm fine with "ugly" merge commits

Reply:
Unfortunately we'd still need to ensure that the dynamically allocated sequence wasn't free'd before the sequence had finished playing. The fix in flipperdevices/flipperzero-firmware#2335 fixes the case where the sequence is defined statically. What would happen was the program would be unloaded from memory, but the sequence still had a pointer to somewhere in the now free'd memory region. The fix just makes sure the Flipper runtime waits until all previous sequences have finished playing before unloading the program from memory.

The problem for a dynamically allocated sequences is that all this happens **after** main has exited:

```rust
// main
{
    // create dynamically allocated sequence
    let s = vec![...];
    
    play(s); // non-blocking
    
    // s is dropped, causing sequence to be freed
}

// oops! The flipper runtime wasn't finished with that sequence!
// CRASH!

wait_for_all_sequences_to_finish();
```

Reply:
#49 will result in apps waiting for un-joined threads to finish before exiting the FAP, so I think a similar wait here would be fine (assuming the SDK provides a way for us to determine when it has finished).

Reply:
I think the eventual answer here will need to be that `play(s)` takes dynamic sequences by value, and under the hood stores them in an app-static reference (using e.g. #64) to ensure their lifetime extends beyond the user's `main()`. Then we use the same waiting technique that the firmware does, but inside `entry!()` instead (like #49 does for threads).